File: 프로그래머스/3/12942. 최적의 행렬 곱셈/최적의 행렬 곱셈.cpp
#include <string>
#include <vector>

using namespace std;
vector<vector<int>> matrix;
int dp[200][200];
int go(int left, int right){
    if(left>=right){
        return 0;
    }
    if(dp[left][right]){
        return dp[left][right];
    }
    int res=2*1e9;
    for(int i=left;i<right;i++){
        res=min(res,matrix[left][0]*matrix[i][1]*matrix[right][1]+go(left,i)+go(i+1,right));
    }
    return dp[left][right]=res;
}
int solution(vector<vector<int>> matrix) {
    ::matrix=matrix;
    return go(0,matrix.size()-1);
}

File: 프로그래머스/3/12942. 최적의 행렬 곱셈/최적의 행렬 곱셈.java
class Solution {
    
    int dp[][] = new int[200][200];
    int matrix[][] = new int[200][200];
    public int solution(int[][] matrix) {
        int answer = 0;
        this.matrix=matrix;
        return go(0,matrix.length-1);
    }
    
    int go(int n,int m) {
        if (n>=m){
            return 0;
        }
        if (dp[n][m]>0){
            return dp[n][m];
        }
        int ret=Integer.MAX_VALUE;
        for(int i=n;i<m;i++){
            ret=Math.min(ret,go(n,i)+go(i+1,m)+matrix[n][0]*matrix[i][1]*matrix[m][1]);
        }
        return dp[n][m]=ret;
    }
}